### DISCUSSION TITLE
How to secure HTTP interface?

### DISCUSSION THREAD

**Discussion by ofadam** [gtxaspec/wz_mini_hacks#367] [Q&A]
Is there a way I can secure the HTTP interface of wz_mini_hacks? I'd like to pull still JPGs from the camera, but don't want to open port 80 to the device when the configuration can be edited as well.

**Comment by virmaior** [comment]
Good question. The httpd on the server can't be directly secured, but I was thinking about this very issue and want to add a  config tag to let users pick between four modes of www:
1. full web functionality (edit configs, view configs, car control, image only)
2. no edits (view configs, car control, image)
3. no view of configs (= car control + image)
4. no car control (= image only)

**Reply by ofadam** [reply]
Love your idea of different levels! For me at least, would eliminate the need for actual authentication.

**Comment by KevinKadow** [comment]
Without modifying 'httpd' to handle user authentication?

One approach to restricting the configuration view/edit would be to modify "config.cgi" to add a password field to the form and validate that password before acting on a POST request and before showing the full config.

Note that this would NOT protect the "Reboot" and "Revert" action, as those are GET requests, nor car control, etc.

**Comment by virmaior** [comment]
pull request is in to add
`WEB_SERVER_OPTIONS="cam config car jpeg"`
to `wz_mini.conf`
which would then allow you to pick what is exposed on the web interface.

@KevinKadow without cookies constantly posting the password is a bit of a nightmare to maintain. ash (not exactly bash) is not really designed to do cgi web serving.

**Reply by mrlt8** [reply]
What about something like an api key? It could be something like a hash of the camera's mac address e.g. `/cgi-bin/jpeg.cgi?api=4a7074688ea4a32cd257f7c672d56365`

**Reply by virmaior** [reply]
that's a bit of a secret code solution but not out of the question to implement with the resources we have. It could a hash of the camera's mac or even a stored value. 
Similar idea would be to have some sort of base64 link to unlock where the user can set that if they want that sort of security-through-obfuscation (which doesn't greatly concern me considering I *really hope no one is using a wyze camera near anything that needs real security*).

**Comment by endertable** [comment]
@ofadam If you are talking about SSL security, there is no binary on Wyze firmwares.  The lightest 3rd party available is lighttpd.  I have been using it for short of a few years now.  If you would like to try it then grab the attached file, download it to the ROOT of the SD card and issue the following (it is provisioned for wz_mini but you can change whatever):

```
gzip -d wz_mini_lighttpd.tar.gz
tar xvf wz_mini_lighttpd.tar
./lh_install
```

Enter your new web password twice (the username is 'root' and I wouldn't change that unless you know what you are doing), and then it will make lighttpd your new default web server then it will attempt to start the web server (the first time, it will create a 2 year certificate file.  Self-signed of course).  Logs are saved in the usual place.  For anyone who tries it, let me know how it works, or not.


[wz_mini_lighttpd.tar.gz](https://github.com/gtxaspec/wz_mini_hacks/files/9893106/wz_mini_lighttpd.tar.gz)
